### PR TITLE
Add an experimental pub-sub API

### PR DESCRIPTION
**Pull request by bschwind**
This is just to get a discussion going, there are many obvious flaws with this approach, but it has the start of something good. Let's iterate on it! It feels like a decent API to start with.

### CODE REVIEW

**Review by strohel — COMMENTED** [review]
@bschwind finally got to have a look. This is indeed nice and more straightforward than I thought!

Perhaps `subscribe()` could be on `Context` rather than on `SpawnBuilder`? It could be most natural to call it from the actor itself.

We could also have `Event` marker trait that we could require being implemented for events, a kind of extra type safety.

I'm also thinking whether it would be possible for actors to somehow statically subscribe to events. I need to study how bevy does the magic (where you just mention it as a parameter of a function and it magically works).

**Comment by bschwind** [comment]
@strohel good idea! I moved the `subscribe()` function to `Context` and I think that works nicely. Take a look at the example usage and let me know what you think.

**Comment by skywhale** [comment]
Let's quickly iterate on this together once @strohel is in Tokyo :)

File: examples/pub_sub_example.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        Self { counter: 0, started_at: Instant::now() }
+    }
+}
+
+impl Actor for PublisherActor {
+    type Context = Context<Self::Message>;
+    type Error = Error;
+    type Message = PublisherMessage;
+
+    fn name() -> &'static str {
```
Comment (PabloMansanet) [inline examples/pub_sub_example.rs:34]: Off-topic, but I wonder if we can have a default implementation of this method using [type_name](https://doc.rust-lang.org/std/any/fn.type_name.html), since it's mostly what we're manually doing anyway.

File: examples/pub_sub_example.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        context.set_deadline(Some(deadline + Duration::from_secs(1)));
+        Ok(())
+    }
+}
+
+enum SubscriberMessage {
+    Text(String),
+}
+
+impl From<String> for SubscriberMessage {
```
Comment (PabloMansanet) [inline examples/pub_sub_example.rs:85]: I appreciate using `String` as the "topic" type for simplicity on the example, but I feel like it made it a bit hard to wrap my head around it (It took me a bit to realize that `String` was the topic being subscribed to). 

Maybe a newtype over String would help for the example? `struct Topic(String)`

File: src/lib.rs [hunk 3]
```diff
@@ -236,6 +246,24 @@ impl<M> Context<M> {
     pub fn set_timeout(&mut self, timeout: Option<Duration>) {
         self.set_deadline(timeout.map(|t| Instant::now() + t));
     }
+
+    pub fn subscribe<E: 'static + Into<M> + Clone>(&mut self)
```
Comment (PabloMansanet) [inline src/lib.rs:250]: Smart trait bounds :) I like it.

File: src/lib.rs [hunk 4]
```diff
@@ -236,6 +246,24 @@ impl<M> Context<M> {
...
+    where
+        M: 'static,
+    {
+        let mut event_subscribers = self.system_handle.event_subscribers.lock();
+        let event_addr = self.myself.clone();
+
+        // TODO(bschwind) - panic if the events HashMap doesn't contain this event type ID.
+        event_subscribers.events.entry(TypeId::of::<E>()).and_modify(|subs| {
+            subs.push(Box::new(move |e| {
```
Comment (PabloMansanet) [inline src/lib.rs:259]: These nested closures are a bit hard to parse visually. How about extracting the inner one into a standalone variable with a descriptive name? (Right before the call to `entry`)

File: src/lib.rs [hunk 5]
```diff
@@ -645,6 +682,15 @@ impl SystemHandle {
         }
     }
 
+    pub fn publish<E: 'static + std::any::Any>(&self, event: E) {
+        let event_subscribers = self.event_subscribers.lock();
+        if let Some(subs) = event_subscribers.events.get(&TypeId::of::<E>()) {
+            for sub in subs {
+                sub(&event);
```
Comment (PabloMansanet) [inline src/lib.rs:689]: The one thing I don't like about this is that it could get very costly (imagine a big list of subscribers with a big event type being cloned and converted into their respective message types), and this cost is paid upfront by the caller to `publish`. 

It feels like for the general use case of a pub/sub framework, a high priority thread that's doing the publishing won't be able to bear this cost, so it seems like there's an abstraction missing in the middle (some sort of queue that takes these published events and distributes them to the subscribers). This middleman then could be prioritized appropriately so it doesn't slow down the high priority thread that just needs to publish and keep going.

Comment (PabloMansanet) [reply src/lib.rs:689]: After local conversation:

It may be interesting to, instead of sending the whole topic struct, instead send a reference counted handle to the same heap data to all of the subscribers, which is a lot less costly and would alleviate most of my issues with this (though I'm still a bit worried about deadlocks).

File: src/lib.rs [hunk 6]
```diff
@@ -201,6 +213,8 @@ pub struct SystemHandle {
     registry: Arc<Mutex<Vec<RegistryEntry>>>,
     system_state: Arc<RwLock<SystemState>>,
     callbacks: Arc<SystemCallbacks>,
+
+    event_subscribers: Arc<Mutex<EventSubscribers>>,
```
Comment (strohel) [inline src/lib.rs:217]: Now that we have proper benchmarks I wonder whether `RwLock` would be faster for high publisher count. My guess is that it will.

**Review by strohel — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 6 above)
Comment (strohel) [reply src/lib.rs]: It indeed is, pushed a commit that changes to RwLock, bench results:
```
                        time:   [20.233 us 20.472 us 20.730 us]
                        thrpt:  [48.238 Kelem/s 48.847 Kelem/s 49.425 Kelem/s]
                 change:
                        time:   [-2.8148% -1.4032% -0.0071%] (p = 0.06 > 0.05)
                        thrpt:  [+0.0071% +1.4232% +2.8963%]
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
publish/2 publishers 10 subscribers
                        time:   [35.620 us 36.114 us 36.652 us]
                        thrpt:  [27.284 Kelem/s 27.690 Kelem/s 28.074 Kelem/s]
                 change:
                        time:   [-32.864% -31.967% -30.928%] (p = 0.00 < 0.05)
                        thrpt:  [+44.777% +46.988% +48.952%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
publish/10 publishers 10 subscribers
                        time:   [13.782 us 14.184 us 14.577 us]
                        thrpt:  [68.602 Kelem/s 70.503 Kelem/s 72.561 Kelem/s]
                 change:
                        time:   [-94.001% -93.697% -93.410%] (p = 0.00 < 0.05)
                        thrpt:  [+1417.4% +1486.4% +1566.9%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high severe
publish/1 publishers 100 subscribers
                        time:   [237.29 us 238.86 us 240.37 us]
                        thrpt:  [4.1602 Kelem/s 4.1865 Kelem/s 4.2142 Kelem/s]
                 change:
                        time:   [-2.2219% -1.2377% -0.2828%] (p = 0.02 < 0.05)
                        thrpt:  [+0.2836% +1.2533% +2.2724%]
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) low mild
  1 (1.00%) high mild
publish/2 publishers 100 subscribers
                        time:   [379.46 us 383.71 us 388.60 us]
                        thrpt:  [2.5734 Kelem/s 2.6062 Kelem/s 2.6353 Kelem/s]
                 change:
                        time:   [-22.200% -20.768% -19.282%] (p = 0.00 < 0.05)
                        thrpt:  [+23.887% +26.211% +28.535%]
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
publish/10 publishers 100 subscribers
                        time:   [1.3161 ms 1.3673 ms 1.4093 ms]
                        thrpt:  [709.56  elem/s 731.35  elem/s 759.84  elem/s]
                 change:
                        time:   [-48.118% -43.801% -39.608%] (p = 0.00 < 0.05)
                        thrpt:  [+65.586% +77.938% +92.744%]
                        Performance has improved.
```

File: src/lib.rs (hunk 5 above)
Comment (bschwind) [reply src/lib.rs]: @strohel and I discussed this a bit, and our conclusion was that the decision is probably better left to the users of the API, so they can wrap their message in an `Arc`, or not if they don't mind the clones.

File: src/lib.rs (hunk 6 above)
Comment (bschwind) [reply src/lib.rs]: Nice one! Good performance gains for very little effort, I like it :)

Comment (bschwind) [reply src/lib.rs]: Here are benchmarks from my machine:

```
publish/1 publishers 10 subscribers                        
                        time:   [4.9240 us 5.3191 us 5.8158 us]
                        thrpt:  [171.95 Kelem/s 188.00 Kelem/s 203.09 Kelem/s]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
Benchmarking publish/2 publishers 10 subscribers: Collecting 100 samples in estimated 5.0471                                                                                              publish/2 publishers 10 subscribers                        
                        time:   [18.367 us 19.405 us 20.391 us]
                        thrpt:  [49.042 Kelem/s 51.533 Kelem/s 54.445 Kelem/s]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) low mild
  1 (1.00%) high mild
Benchmarking publish/10 publishers 10 subscribers: Collecting 100 samples in estimated 5.0423                                                                                             publish/10 publishers 10 subscribers                        
                        time:   [111.58 us 115.57 us 120.38 us]
                        thrpt:  [8.3072 Kelem/s 8.6524 Kelem/s 8.9619 Kelem/s]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe
Benchmarking publish/1 publishers 100 subscribers: Collecting 100 samples in estimated 5.6701                                                                                             publish/1 publishers 100 subscribers                        
                        time:   [128.32 us 140.51 us 153.48 us]
                        thrpt:  [6.5156 Kelem/s 7.1170 Kelem/s 7.7931 Kelem/s]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  7 (7.00%) high mild
Benchmarking publish/2 publishers 100 subscribers: Collecting 100 samples in estimated 5.5443                                                                                             publish/2 publishers 100 subscribers                        
                        time:   [141.33 us 171.77 us 204.92 us]
                        thrpt:  [4.8799 Kelem/s 5.8218 Kelem/s 7.0755 Kelem/s]
Found 14 outliers among 100 measurements (14.00%)
  6 (6.00%) high mild
  8 (8.00%) high severe
Benchmarking publish/10 publishers 100 subscribers: Collecting 100 samples in estimated 5.673                                                                                             publish/10 publishers 100 subscribers                        
                        time:   [1.0031 ms 1.1075 ms 1.2075 ms]
                        thrpt:  [828.14  elem/s 902.96  elem/s 996.92  elem/s]
Found 21 outliers among 100 measurements (21.00%)
  8 (8.00%) low severe
  2 (2.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe
```

File: src/lib.rs (hunk 5 above)
Comment (strohel) [reply src/lib.rs]: I've also realized that forcing `Arc` would be limiting: it requires `T` to be `Sync` for `Arc<T>` to be `Send`, whereas currently we only require `Event` to be `Send` and not necessarily `Sync`.

File: examples/pub_sub_example.rs (hunk 1 above)
Comment (strohel) [reply examples/pub_sub_example.rs:39]: @PabloMansanet good idea, I've filed #68; resolving this thread.

File: src/lib.rs (hunk 4 above)
Comment (strohel) [reply src/lib.rs]: We've now removed one level of closure by using `Entry::or_default()`, so this can be hopefully considered resolved.

**Comment by strohel** [comment]
### Benchmark results for https://github.com/tonarino/actor/commit/90089c15e39bc366f66a6fbb320ade5c40e15abd (RwLock)

#### Thinkpad T480 4 Intel cores:
```
publish/1 publishers 10 subscribers
                        time:   [20.233 us 20.472 us 20.730 us]
                        thrpt:  [48.238 Kelem/s 48.847 Kelem/s 49.425 Kelem/s]
                 change:
                        time:   [-2.8148% -1.4032% -0.0071%] (p = 0.06 > 0.05)
                        thrpt:  [+0.0071% +1.4232% +2.8963%]
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
publish/2 publishers 10 subscribers
                        time:   [35.620 us 36.114 us 36.652 us]
                        thrpt:  [27.284 Kelem/s 27.690 Kelem/s 28.074 Kelem/s]
                 change:
                        time:   [-32.864% -31.967% -30.928%] (p = 0.00 < 0.05)
                        thrpt:  [+44.777% +46.988% +48.952%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
publish/10 publishers 10 subscribers
                        time:   [13.782 us 14.184 us 14.577 us]
                        thrpt:  [68.602 Kelem/s 70.503 Kelem/s 72.561 Kelem/s]
                 change:
                        time:   [-94.001% -93.697% -93.410%] (p = 0.00 < 0.05)
                        thrpt:  [+1417.4% +1486.4% +1566.9%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high severe
publish/1 publishers 100 subscribers
                        time:   [237.29 us 238.86 us 240.37 us]
                        thrpt:  [4.1602 Kelem/s 4.1865 Kelem/s 4.2142 Kelem/s]
                 change:
                        time:   [-2.2219% -1.2377% -0.2828%] (p = 0.02 < 0.05)
                        thrpt:  [+0.2836% +1.2533% +2.2724%]
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) low mild
  1 (1.00%) high mild
publish/2 publishers 100 subscribers
                        time:   [379.46 us 383.71 us 388.60 us]
                        thrpt:  [2.5734 Kelem/s 2.6062 Kelem/s 2.6353 Kelem/s]
                 change:
                        time:   [-22.200% -20.768% -19.282%] (p = 0.00 < 0.05)
                        thrpt:  [+23.887% +26.211% +28.535%]
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
publish/10 publishers 100 subscribers
                        time:   [1.3161 ms 1.3673 ms 1.4093 ms]
                        thrpt:  [709.56  elem/s 731.35  elem/s 759.84  elem/s]
                 change:
                        time:   [-48.118% -43.801% -39.608%] (p = 0.00 < 0.05)
                        thrpt:  [+65.586% +77.938% +92.744%]
                        Performance has improved.
```

#### Apple MacBook Pro 2021 10-core M1 Pro
```
publish/1 publishers 10 subscribers                        
                        time:   [4.9240 us 5.3191 us 5.8158 us]
                        thrpt:  [171.95 Kelem/s 188.00 Kelem/s 203.09 Kelem/s]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
Benchmarking publish/2 publishers 10 subscribers: Collecting 100 samples in estimated 5.0471                                                                                              publish/2 publishers 10 subscribers                        
                        time:   [18.367 us 19.405 us 20.391 us]
                        thrpt:  [49.042 Kelem/s 51.533 Kelem/s 54.445 Kelem/s]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) low mild
  1 (1.00%) high mild
Benchmarking publish/10 publishers 10 subscribers: Collecting 100 samples in estimated 5.0423                                                                                             publish/10 publishers 10 subscribers                        
                        time:   [111.58 us 115.57 us 120.38 us]
                        thrpt:  [8.3072 Kelem/s 8.6524 Kelem/s 8.9619 Kelem/s]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe
Benchmarking publish/1 publishers 100 subscribers: Collecting 100 samples in estimated 5.6701                                                                                             publish/1 publishers 100 subscribers                        
                        time:   [128.32 us 140.51 us 153.48 us]
                        thrpt:  [6.5156 Kelem/s 7.1170 Kelem/s 7.7931 Kelem/s]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  7 (7.00%) high mild
Benchmarking publish/2 publishers 100 subscribers: Collecting 100 samples in estimated 5.5443                                                                                             publish/2 publishers 100 subscribers                        
                        time:   [141.33 us 171.77 us 204.92 us]
                        thrpt:  [4.8799 Kelem/s 5.8218 Kelem/s 7.0755 Kelem/s]
Found 14 outliers among 100 measurements (14.00%)
  6 (6.00%) high mild
  8 (8.00%) high severe
Benchmarking publish/10 publishers 100 subscribers: Collecting 100 samples in estimated 5.673                                                                                             publish/10 publishers 100 subscribers                        
                        time:   [1.0031 ms 1.1075 ms 1.2075 ms]
                        thrpt:  [828.14  elem/s 902.96  elem/s 996.92  elem/s]
Found 21 outliers among 100 measurements (21.00%)
  8 (8.00%) low severe
  2 (2.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe
```

#### dev-annex 8-core AMD Threadripper

```
publish/1 publishers 10 subscribers                                                                             
                        time:   [6.3256 us 6.4633 us 6.6055 us]
                        thrpt:  [151.39 Kelem/s 154.72 Kelem/s 158.09 Kelem/s]
publish/2 publishers 10 subscribers                                                                             
                        time:   [9.1439 us 9.2688 us 9.3967 us]
                        thrpt:  [106.42 Kelem/s 107.89 Kelem/s 109.36 Kelem/s]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
publish/10 publishers 10 subscribers                                                                             
                        time:   [3.1484 us 3.1792 us 3.2111 us]
                        thrpt:  [311.42 Kelem/s 314.55 Kelem/s 317.62 Kelem/s]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  2 (2.00%) high severe
publish/1 publishers 100 subscribers                                                                            
                        time:   [61.651 us 62.004 us 62.379 us]
                        thrpt:  [16.031 Kelem/s 16.128 Kelem/s 16.220 Kelem/s]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
publish/2 publishers 100 subscribers                                                                             
                        time:   [88.821 us 89.260 us 89.852 us]
                        thrpt:  [11.129 Kelem/s 11.203 Kelem/s 11.259 Kelem/s]
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) high mild
  10 (10.00%) high severe
publish/10 publishers 100 subscribers                                                                             
                        time:   [309.70 us 311.44 us 312.88 us]
                        thrpt:  [3.1961 Kelem/s 3.2109 Kelem/s 3.2289 Kelem/s]
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) low mild
```

**Review by strohel — APPROVED** [review]
I'm good with the latest version.

I'm now co-author of the PR, so would be cool if somebody else could also approve,  perhaps @PabloMansanet. @bschwind please also have a look at the last 8 commits that I've pushed since we've paired on this.

**Review by bschwind — COMMENTED** [review]
@strohel I reviewed your latest commits, everything is looking good to me! Thanks for taking on the last part of the PR, it's a lot more polished than it started now :)

File: src/lib.rs [hunk 7]
```diff
@@ -645,6 +686,37 @@ impl SystemHandle {
...
+                recipient.send(msg.into())?;
+            }
+
+            Ok(())
+        }));
+    }
+
+    /// Publish an event. All actors that have previously subscribed to the type will receive it.
+    ///
+    /// When sending to some subscriber fails, this method returns an error immediately.
```
Comment (PabloMansanet) [inline src/lib.rs:707]: Is this always desirable? I imagine you'll likely always want to panic if this happens, but I'm not sure if I like that this will deliver the publish to all receivers before the erroring one but to none after it. If someone tries to recover gracefully from this error, they might have a hard time knowing the state of the system precisely.

**Review by PabloMansanet — APPROVED** [review]
Looking great! Not much to add since last time, and since this will never be no_std I'm not that worried anymore about priority inversion :+1:

**Review by bschwind — COMMENTED** [review]
LGTM++
